python/ray/data/tests/test_dataset_formats.py

```diff
@@ -1457,6 +1457,27 @@ def test_csv_write_block_path_provider(
     assert df.equals(ds_df)
 
 
+def test_torch_datasource(ray_start_regular_shared, local_path):
+    import torchvision
+
+    # Download datasets to separate folders to prevent interference.
+    torch_dataset_root = os.path.join(local_path, "torch")
+    ray_dataset_root = os.path.join(local_path, "ray")
+
+    torch_dataset = torchvision.datasets.MNIST(torch_dataset_root, download=True)
+    expected_data = list(torch_dataset)
+
+    def dataset_factory():
+        return torchvision.datasets.MNIST(ray_dataset_root, download=True)
+
+    ray_dataset = ray.data.read_datasource(
+        SimpleTorchDatasource(), parallelism=1, dataset_factory=dataset_factory
+    )
+    actual_data = list(next(ray_dataset.iter_batches(batch_size=ray_dataset.count())))
+
+    assert actual_data == expected_data
```

Comment:
Could you also check that passing an instantiated Torch dataset raises an error? E.g.
```python
with pytest.raises(ValueError):
    ray.data.read_datasource(
        SimpleTorchDatasource,
        parallelism=1,
        dataset_factory=dataset_factory(),
    )
```

Reply:
Fixed